verilog/analysis/checkers/forbid_implicit_declarations_rule_test.cc

```diff
@@ -0,0 +1,54 @@
+// Copyright 2017-2020 The Verible Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "verilog/analysis/checkers/forbid_implicit_declarations_rule.h"
+
+#include <initializer_list>
+
+#include "gtest/gtest.h"
+#include "common/analysis/linter_test_utils.h"
+#include "common/analysis/syntax_tree_linter_test_utils.h"
+#include "common/text/symbol.h"
+#include "verilog/CST/verilog_nonterminals.h"
+#include "verilog/CST/verilog_treebuilder_utils.h"
+#include "verilog/analysis/verilog_analyzer.h"
+
+namespace verilog {
+namespace analysis {
+namespace {
+
+using verible::LintTestCase;
+using verible::RunLintTestCases;
+
+TEST(ForbidImplicitDeclarationsRule, FunctionFailures) {
+  auto kToken = SymbolIdentifier;
+  const std::initializer_list<LintTestCase> ForbidImplicitDeclarationsTestCases = {
```

Comment:
No, an implicit net cannot be created by referencing them on the RHS. Of course, this is according to the LRM, vendor's implementation can differ. 

Using implicit declaration with array (`assign a[0] = 1'b1`) is also illegal. I believe the array indices must be resolved before the assignment, therefore if `a` does not exist, this would fail.

Another missing test case is implicit net declared in a terminal/port connection list:

```
module m;
    wire a;
    not U1(implicit_net_0, a);
    my_m U2(implicit_net_1);
endmodule
```

Reply:
Great thanks for the clarification, and thanks for suggesting more test cases!
Your test case is exactly the situation I was thinking of (not usage in an expression but rather in a port connection).  To that, I would add:

```systemverilog
module m;
  my_m U3(.foo(implicit_net_2));
  my_m U4(.nonexistent_net_in_this_scope);   // assuming .x is equiv. to .x(x)
endmodule
```